sky/serve/serve_state.py

```diff
@@ -382,3 +411,83 @@ def total_number_provisioning_replicas() -> int:
         if replica_info.status == ReplicaStatus.PROVISIONING:
             provisioning_count += 1
     return provisioning_count
+
+
+# === Version functions ===
+def add_version(service_name: str) -> int:
+    """Adds a version to the database."""
+
+    _DB.cursor.execute(
+        """\
+        SELECT MAX(version) FROM versions
+        WHERE service_name=(?)""", (service_name,))
+
+    # add_version is only called during update.
+    new_version = _DB.cursor.fetchone()[0] + 1
+
+    _DB.cursor.execute(
+        """\
+        INSERT INTO versions
+        (version, service_name, spec)
+        VALUES (?, ?, ?)""", (new_version, service_name, pickle.dumps(None)))
+    _DB.conn.commit()
+
+    return new_version
+
+
+def add_or_update_version(service_name: str, version: int,
+                          spec: 'service_spec.SkyServiceSpec') -> None:
+    # Check if the entry with the specified service_name and version exists
+    _DB.cursor.execute(
+        """SELECT * FROM versions WHERE service_name=? AND version=?""",
+        (service_name, version))
+    existing_entry = _DB.cursor.fetchone()
+
+    if existing_entry:
+        _DB.cursor.execute(
+            """\
+            UPDATE versions SET spec=?
+            WHERE service_name=? AND version=?""", (
+                pickle.dumps(spec),
+                service_name,
+                version,
+            ))
+    else:
+        _DB.cursor.execute(
+            """\
+            INSERT INTO versions
+            (service_name, version, spec)
+            VALUES (?, ?, ?)""", (service_name, version, pickle.dumps(spec)))
+    _DB.conn.commit()
```

Comment:
Can we use `INSERT or REPLACE`? https://github.com/skypilot-org/skypilot/blob/f78a16e254faeb3667916d22d55b073339152d09/sky/global_user_state.py#L189

Reply:
Actually, if we always call `add_version` before add the specs, why should we have the `INSERT` part?